File: _CodeTopics/LeetCode/401-600/000451/tran/000451.cpp
class Solution {
public:
    unordered_map<char, int> str_to_dic(string s) {
        unordered_map<char, int> res;
        for (char ch : s) {
            if (res.find(ch) != res.end()) {
                res[ch] += 1;
            } else {
                res[ch] = 1;
            }
        }
        return res;
    }

    // C++ unordered_map 没有类似 Python 的 .keys(), .values() 接口。。。只能自己实现- -
    vector<char> get_list_of_keys(unordered_map<char, int> ump) {
        vector<char> res;
        for (auto kv : ump) {
            res.push_back(kv.first);
        }
        return res;
    }

    string frequencySort(string s) {
        unordered_map<char, int> dic = str_to_dic(s);
        vector<char> keys = get_list_of_keys(dic);

        // 匿名函数想要用外部变量dic，需要中括号里加 &。
        sort(keys.begin(), keys.end(), 
            [&](char x, char y) {
                return (dic[x] >= dic[y]);
            }
        );

        string res = (string)"";
        for (char key : keys) {
            for (int i = 0; i < dic[key]; ++i) {
                res += key;
            }
        }
        return res;
    }
};

/*
https://leetcode-cn.com/submissions/detail/191873952/

32 / 32 个通过测试用例
状态：通过
执行用时: 20 ms
内存消耗: 8.6 MB

执行用时：20 ms, 在所有 C++ 提交中击败了24.46%的用户
内存消耗：8.6 MB, 在所有 C++ 提交中击败了46.96%的用户
*/

File: _CodeTopics/LeetCode/401-600/000451/000451.py
class Solution(object):
    def frequencySort(self, s):
        """
        :type s: str
        :rtype: str
        """

        def str_to_dict_v2(s):
            res = {}
            for ch in s:
                res[ch] = res.setdefault(ch, 0) + 1
            return res
        
        dic = str_to_dict_v2(s)
        keys = dic.keys()
        keys.sort(key = lambda x : dic[x], reverse=True)
        
        s = ''
        for key in keys:
            s += key * dic[key]
        return s
        
"""
https://leetcode-cn.com/submissions/detail/191861163/

32 / 32 个通过测试用例
状态：通过
执行用时: 40 ms
内存消耗: 14.9 MB

执行用时：40 ms, 在所有 Python 提交中击败了87.01%的用户
内存消耗：14.9 MB, 在所有 Python 提交中击败了68.36%的用户
"""
